src/processout/invoice-action.ts

```diff
@@ -0,0 +1,149 @@
+/// <reference path="../references.ts" />
+
+module ProcessOut {
+  /**
+   * Configuration for the invoice action operation.
+   */
+  export type InvoiceActionProps = {
+    /**
+     * Invoice ID.
+     * @var {string}
+     */
+    invoiceId: string;
+    /**
+     * Gateway Configuration to be used for the invoice action.
+     * @var {string}
+     */
+    gatewayConf: InvoiceGatewayConfigurationProps;
+    /**
+     * Customer Token ID - the ID of the customer token ID to use for the invoice action, if any.
+     * @var {string|undefined}
+     */
+    customerTokenId?: string;
+    /**
+     * Additional data - any additional data that should be passed to the gateway.
+     * @var {Object.<string, string>}
+     */
+    additionalData?: any;
+    /**
+     * Tokenized callback - a function that will be called with the tokenized object created during the invoice action handling.
+     * @var {TokenCallback}
+     */
+    tokenized: TokenCallback;
+    /**
+     * Token error callback - a function that will be called if an error occurs while tokenizing the payment method.
+     * @var {TokenErrorCallback}
+     */
+    tokenError: TokenErrorCallback;
+    /**
+     * Iframe override - an object containing the iframe override configuration.
+     * @var {IframeOverride}
+     */
+    iframeOverride?: IframeOverride;
+  }
+
+  export type InvoiceGatewayConfigurationProps = {
+    /**
+     * Gateway configuration ID.
+     * @var {string}
+     */
+    id: string;
+    /**
+     * Gateway name.
+     * @var {string|undefined}
+     */
+    gatewayName?: string;
+    /**
+     * Gateway logo URL.
+     * @var {string|undefined}
+     */
+    gatewayLogoUrl?: string;
+  }
+
+  export type InvoiceActionUrlProps = {
+    /**
+     * Invoice ID.
+     * @var {string}
+     */
+    invoiceId: string;
+
+    /**
+     * ID of the gateway configuration to be used.
+     * @var {string}
+     */
+    gatewayConfigurationId: string;
+
+    /**
+     * Additional data - any additional data that should be passed to the gateway.
+     * @var {Object.<string, string>}
+     */
+    additionalData: any;
+
+    /**
+     * Customer Token ID - the ID of the customer token ID to use for the invoice action, if any.
+     * @var {string}
+     */
+    customerTokenId?: string;
+  }
+
+  /**
+   * Map gatewayConf parameter to a valid {@link InvoiceGatewayConfigurationProps} instance.
+   * @param {any|string} gatewayConf
+   */
+  export function mapInvoiceGatewayConfigurationProps(gatewayConf?: any): InvoiceGatewayConfigurationProps {
+    if (!gatewayConf) {
+      return null;
+    }
+
+    var gatewayConfigurationId = gatewayConf instanceof String || !gatewayConf.id
```

Comment:
`gatewayConf` parameter can be either an actual GatewayConfiguration object with fields like `id`, `name` or `logo_url`, or just a string with gateway configuration ID. 🤷‍♂️

Reply:
`gatewayConf instanceof String` might not be enough in JS 😄 as it will only detect strings created like this `new String("...")`. To detect strings primitives like `"..."` there is `typeof` 👀 . So something like this:
`gatewayConf instanceof String || typeof gatewayConf === "string"` would be more accurate I think.
That being said the second condition - `!gatewayConf.id` will catch all of it, so maybe using only this one makes most sense? (unless I am missing something 😅)

Reply:
Ah, you're right. Forgot about the `typeof` keyword.

I'm actually thinking of doing this:
```javascript
        var gatewayConfigurationId: string = gatewayConf && gatewayConf.id
      ? gatewayConf.id
      : gatewayConf;;
```

If it's not empty and has `id` property - take it. Auto convert it to `string` otherwise.

Reply:
Done in [607627e](https://github.com/processout/processout.js/pull/111/commits/607627efde1d2052eb8312addc2f60af4b2a3387)